### PR TITLE
feat: implement gateway client

### PR DESCRIPTION
**Pull request by ayeletstarkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)

### CODE REVIEW

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 5 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 5 of 6 files reviewed, all discussions resolved (waiting on @elintul and @MohammadNassar1)

File: crates/gateway/src/gateway_client.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+struct GatewayClient {
+    ip: IpAddr,
+    port: u16,
+    client: Client<HttpConnector>,
+}
+
+// TODO: Use in end_to_end test when merged. Should replace `send_and_verify_transaction`
+// function.
+impl GatewayClient {
+    fn _new(ip: IpAddr, port: u16) -> Self {
```
Comment (MohammadNassar1) [inline crates/gateway/src/gateway_client.rs:25]: Please use pub

File: crates/gateway/src/gateway_client.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+use hyper::{Client, Response};
+use starknet_api::external_transaction::ExternalTransaction;
+
+use crate::errors::GatewayError;
+use crate::starknet_api_test_utils::external_invoke_tx_to_json;
+
+pub type GatewayResult<T> = Result<T, GatewayError>;
+
+// TODO: remove when struct is used in end_to_end test.
+#[allow(dead_code)]
```
Comment (MohammadNassar1) [inline crates/gateway/src/gateway_client.rs:15]: AFAIK, it should be a public struct.

File: crates/gateway/src/gateway_client.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            .method("POST")
+            .uri(format!("http://{}", SocketAddr::from((self.ip, self.port))) + "/add_transaction")
+            .header("content-type", "application/json")
+            .body(Body::from(tx_json))?;
+
+        // Send a POST request with the transaction data as the body
+        let response: Response<Body> = self.client.request(request).await?;
+
+        let response_bytes = response.into_body().collect().await?.to_bytes();
+        let response_string = String::from_utf8(response_bytes.to_vec())?;
```
Comment (MohammadNassar1) [inline crates/gateway/src/gateway_client.rs:41]: Can you convert the response to string directly?

File: crates/gateway/src/gateway_client.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+use crate::errors::GatewayError;
+use crate::starknet_api_test_utils::external_invoke_tx_to_json;
+
+pub type GatewayResult<T> = Result<T, GatewayError>;
+
+// TODO: remove when struct is used in end_to_end test.
+#[allow(dead_code)]
+struct GatewayClient {
+    ip: IpAddr,
```
Comment (MohammadNassar1) [inline crates/gateway/src/gateway_client.rs:17]: Please use SocketAddress.

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
Nice!

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 0 of 12 files reviewed, 4 unresolved discussions (waiting on @ArniStarkware, @elintul, and @MohammadNassar1)

___
*[`crates/gateway/src/gateway_client.rs` line 15 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611512025:-Ny_RLbm5T9rJTiTwrcY:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L15)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

AFAIK, it should be a public struct.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_client.rs` line 17 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611523279:-Ny_S-xB5Ewmi5av6ACQ:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L17)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Please use SocketAddress.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_client.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611511066:-Ny_RC0R6JUx9I0jEYGw:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L25)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Please use pub
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_client.rs` line 41 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611521563:-Ny_S-DFCIbZEmNkvcMI:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L41)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Can you convert the response to string directly?
</blockquote></details>

Done.

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 3 of 11 files at r2, 4 of 7 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 10 of 12 files reviewed, 4 unresolved discussions (waiting on @elintul and @MohammadNassar1)

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 11 of 12 files reviewed, 6 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @MohammadNassar1)

File: crates/gateway/Cargo.toml [hunk 5]
```diff
@@ -16,6 +16,7 @@ axum.workspace = true
 blockifier.workspace = true
 cairo-lang-starknet-classes.workspace = true
 clap.workspace = true
+hyper.workspace = true
```
Comment (MohammadNassar1) [inline crates/gateway/Cargo.toml:19]: Please remove hyper from `dev-dependencies`

**Review by MohammadNassar1 — APPROVED** [review]

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 10 of 12 files reviewed, 7 unresolved discussions (waiting on @ArniStarkware, @elintul, and @MohammadNassar1)

___
*[`crates/gateway/Cargo.toml` line 19 at r5](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611699948:-Ny_iOekAOlE70QW-KAZ:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/f2fd84a39e9bd1718ba4e3b1d73636fca2a03de0/crates/gateway/Cargo.toml#L19)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Please remove hyper from `dev-dependencies`
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_client.rs` line 24 at r5](https://reviewable.io/reviews/starkware-libs/mempool/143#-Ny__zsB244dR1fFohyn:-Ny_iviN495-FmGlEIps:bcg3hm5) ([raw file](https://github.com/starkware-libs/mempool/blob/f2fd84a39e9bd1718ba4e3b1d73636fca2a03de0/crates/gateway/src/gateway_client.rs#L24)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

is `GatewayError` the return type we want? I really don't know.

Gateway error impls Axum or something IIRC.
</blockquote></details>

What other error type would you suggest? It only interacts with the gateway, so I thought that 'GatewayError' would be appropriate.
___
*[`crates/gateway/src/gateway_client.rs` line 34 at r5](https://reviewable.io/reviews/starkware-libs/mempool/143#-Ny__Ei4DN9TOhVw4v_E:-Ny_iZ5V4NaKCNqKZP_p:bdzh32o) ([raw file](https://github.com/starkware-libs/mempool/blob/f2fd84a39e9bd1718ba4e3b1d73636fca2a03de0/crates/gateway/src/gateway_client.rs#L34)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

IIUC, gateway client is not a test util.
If this is indeed the case, please avoid `assert_eq`. Handle the error - return a result if need be.
</blockquote></details>

I think it is a test util. I'll check.

**Review by ArniStarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: 10 of 12 files reviewed, 5 unresolved discussions (waiting on @elintul and @MohammadNassar1)

___
*[`crates/gateway/src/gateway_client.rs` line 24 at r5](https://reviewable.io/reviews/starkware-libs/mempool/143#-Ny__zsB244dR1fFohyn:-NynmCJhEXA5Q6r31Buq:b-4rfhgs) ([raw file](https://github.com/starkware-libs/mempool/blob/f2fd84a39e9bd1718ba4e3b1d73636fca2a03de0/crates/gateway/src/gateway_client.rs#L24)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

What other error type would you suggest? It only interacts with the gateway, so I thought that 'GatewayError' would be appropriate.
</blockquote></details>

Not as important if this is only a test util.
___
*[`crates/gateway/src/gateway_client.rs` line 34 at r5](https://reviewable.io/reviews/starkware-libs/mempool/143#-Ny__Ei4DN9TOhVw4v_E:-Nynm5i1B669MukH6t_U:b-mc3xlh) ([raw file](https://github.com/starkware-libs/mempool/blob/f2fd84a39e9bd1718ba4e3b1d73636fca2a03de0/crates/gateway/src/gateway_client.rs#L34)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

yep, it is a test util
</blockquote></details>

Got it. Please document it somewhere.

**Review by MohammadNassar1 — APPROVED** [review]
> Reviewed 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: all files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware and @elintul)

___
*[`crates/gateway/src/gateway_client.rs` line 41 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611521563:-Nyo97JK1riWXkNUQhez:bcy3mx1) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L41)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see your change.
But that's fine, you can keep it like that.
___
*[`crates/gateway/src/gateway_client.rs` line 26 at r7](https://reviewable.io/reviews/starkware-libs/mempool/143#-Nyo8trpCv3s4XENW2NF:-Nyo8trpCv3s4XENW2NG:bw7sjks) ([raw file](https://github.com/starkware-libs/mempool/blob/55f3c935877f3db6b3c5294823f5b755ab625f3f/crates/gateway/src/gateway_client.rs#L26)):*
> ```XML
> 
>     pub async fn add_tx(&self, tx: ExternalTransaction) -> GatewayResult<String> {
>         let tx_json = external_invoke_tx_to_json(tx);
> ```

Q - Does it work for invoke only?



_Code quote:_
```XML
external_invoke_tx_to_json
```

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul and @MohammadNassar1)

___
*[`crates/gateway/src/gateway_client.rs` line 41 at r4](https://reviewable.io/reviews/starkware-libs/mempool/143#gh-1611521563:-NyoFnjpAAu_ScJy7WYI:bb7m3jc) ([raw file](https://github.com/starkware-libs/mempool/blob/c8d57d11e31ac7662b785b56da8c7cbb9fd890cf/crates/gateway/src/gateway_client.rs#L41)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

I don't see your change.
But that's fine, you can keep it like that.
</blockquote></details>

oops
___
*[`crates/gateway/src/gateway_client.rs` line 26 at r7](https://reviewable.io/reviews/starkware-libs/mempool/143#-Nyo8trpCv3s4XENW2NF:-NyoFM1c31F8HSYj2Ds0:bmtxkqy) ([raw file](https://github.com/starkware-libs/mempool/blob/55f3c935877f3db6b3c5294823f5b755ab625f3f/crates/gateway/src/gateway_client.rs#L26)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Q - Does it work for invoke only?
</blockquote></details>

Currently, yes. I'm working on a more generalized version in a different PR.

**Review by MohammadNassar1 — APPROVED** [review]

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @elintul and @MohammadNassar1)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/143)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul and @MohammadNassar1)

**Pull request merged**